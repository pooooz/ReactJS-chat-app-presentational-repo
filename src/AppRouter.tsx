import React, { FC } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';

import { Sidebar } from './components/Sidebar/Sidebar';
import { Dialogue } from './pages/Dialogues/Dialogue';
import { ChatList } from './pages/Dialogues/components/ChatList/ChatList';
import { Profile } from 'src/pages/Profile/Profile';

export const AppRouter: FC = () => {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Sidebar />}>
          <Route path="profile" element={<Profile />} />
          <Route path="chats">
            <Route index element={<ChatList />} />
            <Route path=":chatId" element={<Dialogue />} />
          </Route>
        </Route>

        <Route path="*" element={<h2 style={{ color: '#00BFFF' }}>404</h2>} />
      </Routes>
    </BrowserRouter>
  );
};
