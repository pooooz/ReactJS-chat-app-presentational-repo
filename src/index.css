body {
  margin: 0;
  font-family: 'Press Start 2P', cursive;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
    background-color: black;
}

#root {
    display: flex;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}
