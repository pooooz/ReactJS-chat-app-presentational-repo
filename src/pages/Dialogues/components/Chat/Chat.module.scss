.dialogue {
  font-family: 'Press Start 2P', cursive;
  display: flex;
  flex-direction: column;
  align-items: center;
  height: 100%;
  flex-grow: 2;
}

.form {
  margin: 20px 0 0 0;
  display: flex;
}

.form__button {
  font-family: 'Press Start 2P', cursive !important;
  background: transparent;
  color: limegreen;
  border: none;
  cursor: pointer;
}

.border {
  border: 1px dashed rgba(46, 66, 7, 0.3);
}