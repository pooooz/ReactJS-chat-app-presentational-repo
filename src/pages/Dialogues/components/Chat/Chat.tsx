import React, { useState, FC } from 'react';
import { Button } from '@mui/material';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { Navigate, useParams } from 'react-router-dom';
import { ThunkDispatch } from 'redux-thunk';

import { MessageListPure } from './components/MessageList/MessageList';
import { WithClasses } from 'src/HOC/WithClasses';
import { Input } from 'src/components/Input/Input';
import { addMessage, addMessageWithReply } from 'src/store/dialogues/slice';
import { DialoguesState } from 'src/store/dialogues/slice';
import { selectChatList, selectChats } from 'src/store/dialogues/selectors';

import styles from './Chat.module.scss';

export const Chat: FC = () => {
  const [inputValue, setInputValue] = useState<string>('');
  const { chatId } = useParams();
  const chats = useSelector(selectChats, shallowEqual);
  const chatList = useSelector(selectChatList, shallowEqual);
  const dispatch =
    useDispatch<
      ThunkDispatch<DialoguesState, void, ReturnType<typeof addMessage>>
    >();
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value);
  };

  const handleSubmit = (event: React.FormEvent) => {
    if (inputValue && chatId) {
      dispatch(
        addMessageWithReply({
          chatId,
          message: { author: 'You', text: inputValue },
        })
      );
      setInputValue('');
    }
    event.preventDefault();
  };

  const MessageListPureWithClass = WithClasses(MessageListPure);

  if (!chatList.find((chat) => chat.name === chatId)) {
    return <Navigate replace to="/chats" />;
  }

  return (
    <section className={styles.dialogue}>
      <MessageListPureWithClass
        messages={chatId ? chats[chatId] : []}
        classes={styles.border}
      />
      <form className={styles.form} onSubmit={handleSubmit}>
        <Input value={inputValue} change={handleChange} type="text" />
        <Button
          className={styles.form__button}
          type="submit"
          style={{
            width: '100px',
            color: '#32CD32FF',
          }}
        >
          Send
        </Button>
      </form>
    </section>
  );
};
