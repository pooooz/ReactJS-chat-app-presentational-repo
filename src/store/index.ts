import { combineReducers } from 'redux';
import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
  REHYDRATE,
  persistReducer,
  persistStore,
} from 'redux-persist';

import { profileReducer } from 'src/store/profile/slice';
import { dialoguesReducer } from 'src/store/dialogues/slice';

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['profile'],
};

export type StoreState = ReturnType<typeof rootReducer>;
const rootReducer = combineReducers({
  profile: profileReducer,
  dialogues: dialoguesReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  devTools: process.env.NODE_ENV !== 'production',
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  reducer: persistedReducer,
});

export const persistor = persistStore(store);
