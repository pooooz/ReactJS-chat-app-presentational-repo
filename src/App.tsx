import React, { FC, useState } from 'react';
import { defaultContext, ThemeContext } from 'src/utils/ThemeContext';
import { Provider } from 'react-redux';

import { AppRouter } from './components/AppRouter';
import { store } from 'src/store';

export const App: FC = () => {
  const [theme, setTheme] = useState(defaultContext.theme);

  const toggleTheme = () => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  };

  return (
    <Provider store={store}>
      <ThemeContext.Provider value={{ theme, toggleTheme }}>
        <AppRouter />
      </ThemeContext.Provider>
    </Provider>
  );
};
