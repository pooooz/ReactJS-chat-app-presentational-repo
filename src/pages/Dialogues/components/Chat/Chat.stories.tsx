import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Provider } from 'react-redux';
import { MemoryRouter, Routes, Route } from 'react-router-dom';

import { Chat } from './Chat';
import { store } from 'src/store';

export default {
  title: 'pages/Dialogues/components/Chat',
  component: Chat,
} as ComponentMeta<typeof Chat>;

export const Default: ComponentStory<typeof Chat> = () => {
  return (
    <Provider store={store}>
      <MemoryRouter initialEntries={['/chats/default']} initialIndex={0}>
        <Routes>
          <Route path="/chats/:chatId" element={<Chat />} />
        </Routes>
      </MemoryRouter>
    </Provider>
  );
};
