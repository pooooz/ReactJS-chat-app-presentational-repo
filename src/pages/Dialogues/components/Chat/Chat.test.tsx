import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import '@testing-library/jest-dom';
import '@testing-library/user-event';

import { Chat } from './Chat';
import { store } from 'src/store';

describe('Chat', () => {
  it('Render component', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/chats/default']} initialIndex={0}>
          <Routes>
            <Route path="/chats/:chatId" element={<Chat />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
  });

  it('Render with snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/chats/default']} initialIndex={0}>
          <Routes>
            <Route path="/chats/:chatId" element={<Chat />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Empty message sending', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/chats/default']} initialIndex={0}>
          <Routes>
            <Route path="/chats/:chatId" element={<Chat />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: '' },
    });
    fireEvent.click(screen.getByRole('button'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('Filled message sending', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/chats/default']} initialIndex={0}>
          <Routes>
            <Route path="/chats/:chatId" element={<Chat />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'Test' },
    });
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByText(/Test/)).toBeInTheDocument();
  });

  it('Wrong route test', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/chats/defaults']} initialIndex={0}>
          <Routes>
            <Route path="/chats/:chatId" element={<Chat />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
  });
});
