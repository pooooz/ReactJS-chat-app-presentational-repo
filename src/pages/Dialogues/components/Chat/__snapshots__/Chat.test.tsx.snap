// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Empty message sending 1`] = `
<DocumentFragment>
  <section
    class="dialogue"
  >
    <div
      class="border"
    >
      <div
        class="message__list"
      >
        <div
          class="message"
        >
          <p
            class="glitch"
            data-text="Initialization..."
          >
            Initialization...
          </p>
          <p
            class="glitch"
            data-text="Author: <Admin>"
          >
            Author: &lt;Admin&gt;
          </p>
        </div>
      </div>
    </div>
    <form
      class="form"
    >
      <div
        class="input_wrap"
      >
        <input
          class="input_green"
          type="text"
          value=""
        />
      </div>
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root form__button css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        style="width: 100px; color: rgba(50, 205, 50, 1);"
        tabindex="0"
        type="submit"
      >
        Send
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </section>
</DocumentFragment>
`;

exports[`Chat Render with snapshot 1`] = `
<DocumentFragment>
  <section
    class="dialogue"
  >
    <div
      class="border"
    >
      <div
        class="message__list"
      >
        <div
          class="message"
        >
          <p
            class="glitch"
            data-text="Initialization..."
          >
            Initialization...
          </p>
          <p
            class="glitch"
            data-text="Author: <Admin>"
          >
            Author: &lt;Admin&gt;
          </p>
        </div>
      </div>
    </div>
    <form
      class="form"
    >
      <div
        class="input_wrap"
      >
        <input
          class="input_green"
          type="text"
          value=""
        />
      </div>
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root form__button css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        style="width: 100px; color: rgba(50, 205, 50, 1);"
        tabindex="0"
        type="submit"
      >
        Send
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </section>
</DocumentFragment>
`;
